Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+
+"""
+We make a DQN agent for playing Atari games, following (and modifying/extended) the tutorial here:
+https://becominghuman.ai/lets-build-an-atari-ai-part-1-dqn-df57e8ff3b26
+
+This code is specifically for Breakout.
+"""
+
+import gym
+import random
+import numpy as np
+import time
+import os
+from tensorflow import keras
+import keras
+from keras import layers
+from keras.models import Model
+
+
+# Make the model
+
+ATARI_SHAPE = (105, 80, 4)  # input image size to model (but note that we actually have to input shape: (batch,105,80,4))
+ACTION_SIZE = 4
+
+def atari_model_mask():
+    # With the functional API we need to define the inputs.
+    frames_input = layers.Input(ATARI_SHAPE, name='frames')
+    actions_input = layers.Input((ACTION_SIZE,), name='action_mask')  # Masking!
+
+    # Assuming that the input frames are still encoded from 0 to 255. Transforming to [0, 1].
+    normalized = layers.Lambda(lambda x: x / 255.0, name='normalization')(frames_input)
```

Comment:
I tend to do this kind of preprocessing before feeding it into the ML model (since we're not computing gradients with stuff like this). In this context I think this approach is totally fine too, but in general what you can do in TensorFlow/Keras/PyTorch is much more restricted than what you can do in Python/NumPy, so best to only use the ML frameworks when you actually need their features (e.g. automatic differentiation or scalability).